### PR TITLE
Fixed for SpringBootPractice3

### PR DESCRIPTION
**Pull request by AnatolyNz**

### CODE REVIEW

File: src/main/java/mate/academy/mapper/BookMapper.java [hunk 1]
```diff
@@ -2,14 +2,25 @@
...
 
     Book toModel(CreateBookRequestDto requestDto);
+
+    Book toEntity(CreateBookRequestDto bookDto);
+
+    BookDtoWithoutCategoryIds toDtoWithoutCategories(Book book);
+
+    @AfterMapping
+    default void setCategoryIds(@MappingTarget BookDto bookDto, Book book) {
+    }
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/mapper/BookMapper.java]: remove or implement

File: src/main/java/mate/academy/mapper/OrderItemMapper.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import mate.academy.dto.OrderItemResponseDto;
+import mate.academy.model.CartItem;
+import mate.academy.model.OrderItem;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class)
+public interface OrderItemMapper {
+    @Mapping(target = "book", source = "book")
+    @Mapping(target = "quantity", source = "quantity")
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/mapper/OrderItemMapper.java]:
```suggestion
```

File: src/main/java/mate/academy/mapper/OrderMapper.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+    @Mapping(target = "userId", source = "user.id")
+    @Mapping(target = "orderStatus", source = "status")
+    OrderResponseDto toDto(Order order);
+
+    Order toModel(OrderRequestDto orderRequestDto);
+
+    @Mapping(target = "userId", source = "user.id")
+    @Mapping(target = "orderItems", source = "orderItems")
+    @Mapping(target = "total", source = "total")
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/mapper/OrderMapper.java]:
```suggestion
```

File: src/main/java/mate/academy/model/OrderItem.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+@Where(clause = "is_deleted=false")
+@Accessors(chain = true)
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @JoinColumn(name = "order_id", nullable = false)
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/model/OrderItem.java]: use
@EqualsAndHashCode(exclude = {})
@ToString(exclude = {})
on class level

File: src/main/java/mate/academy/service/CategoryServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+@Service
+public class CategoryServiceImpl implements CategoryService {
+    private final CategoryRepository categoryRepository;
+    private final CategoryMapper categoryMapper;
+
+    @Override
+    public List<CategoryResponseDto> findAll(Pageable pageable) {
+        return categoryRepository.findAll(pageable).stream()
+                .map(categoryMapper::toDto)
+                .collect(Collectors.toList());
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/service/CategoryServiceImpl.java]: Mapstruct supports mapping collections from the box

File: src/main/java/mate/academy/service/CategoryServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Override
+    public CategoryResponseDto save(CategoryDto categoryDto) {
+        Category category = categoryMapper.toEntity(categoryDto);
+        Category savedCategory = categoryRepository.save(category);
+        return categoryMapper.toDto(savedCategory);
+    }
+
+    @Override
+    public CategoryResponseDto update(Long id, CategoryDto categoryDto) {
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/service/CategoryServiceImpl.java:43]: Transaction

File: src/main/java/mate/academy/service/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        double total = shoppingCart.getCartItems().stream()
+                .mapToDouble(cartItem -> (double) cartItem.getQuantity()
+                        * cartItem.getBook().getPrice().doubleValue())
+                .sum();
+        Order order = new Order();
+        order.setUser(user);
+        order.setShippingAddress(shippingAddress.getShippingAddress());
+        order.setOrderDate(LocalDateTime.now());
+        order.setStatus(Order.Status.NEW);
+        order.setTotal(BigDecimal.valueOf(total));
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/service/OrderServiceImpl.java]: use MapStruct

File: src/main/java/mate/academy/service/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        Set<OrderItem> orderItems = shoppingCart.getCartItems().stream()
+                .map(cartItem -> {
+                    OrderItem orderItem = new OrderItem();
+                    orderItem.setBook(cartItem.getBook());
+                    orderItem.setQuantity(cartItem.getQuantity());
+                    orderItem.setPrice(cartItem.getBook().getPrice());
+                    orderItem.setOrder(order);
+                    return orderItemRepository.save(orderItem);
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/service/OrderServiceImpl.java]: mapsctruct

File: src/main/java/mate/academy/service/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        return orderRepository.save(order);
+    }
+
+    @Override
+    public List<OrderResponseDto> getAllOrders(User user,
+                                               Pageable pageable) {
+        Page<Order> allOrders = orderRepository.findAllByUserId(user.getId(), pageable);
+        return allOrders.stream()
+                .map(orderMapper::toResponseDto)
+                .toList();
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/service/OrderServiceImpl.java]: mapstctrct

File: src/main/java/mate/academy/service/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        return orderMapper.toDto(orderRepository.save(order));
+    }
+
+    @Override
+    public List<OrderItemResponseDto> getAllOrderItems(User user, Long orderId, Pageable pageable) {
+        Order order = orderRepository.findByUserIdAndId(user.getId(), orderId)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find order by order id "
+                        + orderId + " and user id " + user.getId()));
+        return (List<OrderItemResponseDto>) orderItemMapper
+                .toResponseDto((OrderItem) order.getOrderItems());
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/service/OrderServiceImpl.java]: why we need casting?

File: src/main/java/mate/academy/service/ShoppingCartServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(cartItemId,
+                        shoppingCart.getId())
+                .orElseThrow(() -> new EntityNotFoundException("Can't find cart item "
+                        + "by cart item id " + cartItemId));
+        cartItem.setQuantity(qtyRequestDto.getQuantity());
+        cartItemRepository.save(cartItem);
+    }
+
+    private ShoppingCart registerNewShoppingCart(User user) {
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/service/ShoppingCartServiceImpl.java]: private methods after public ones

**Review by MAMentorFX — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/service/OrderServiceImpl.java (hunk 10 above)
Comment (AnatolyNz) [reply src/main/java/mate/academy/service/OrderServiceImpl.java]: Without casting I have error.

File: src/main/java/mate/academy/service/OrderServiceImpl.java [hunk 12]
```diff
@@ -44,19 +44,12 @@ public Order placeOrder(User user, ShippingAddressRequestDto shippingAddress) {
...
                         * cartItem.getBook().getPrice().doubleValue())
                 .sum();
-        Order order = new Order();
-        order.setUser(user);
-        order.setShippingAddress(shippingAddress.getShippingAddress());
-        order.setOrderDate(LocalDateTime.now());
-        order.setStatus(Order.Status.NEW);
-        order.setTotal(BigDecimal.valueOf(total));
+        OrderRequestDto orderRequestDto = orderMapper.toDto(shoppingCart);
+        Order order = orderMapper.toModel(orderRequestDto);
```
Comment (olekskov) [inline src/main/java/mate/academy/service/OrderServiceImpl.java]: I cannot see reason why shopping cart should be mapped to OrderRequestDto to instantly remapped to Order

just make constructor for Order that accepts ShoppingCart
```suggestion
        Order order = new Order(shoppingCart);
```

File: src/main/java/mate/academy/service/OrderServiceImpl.java [hunk 13]
```diff
@@ -44,19 +44,12 @@ public Order placeOrder(User user, ShippingAddressRequestDto shippingAddress) {
...
         Set<OrderItem> orderItems = shoppingCart.getCartItems().stream()
                 .map(cartItem -> {
-                    OrderItem orderItem = new OrderItem();
-                    orderItem.setBook(cartItem.getBook());
-                    orderItem.setQuantity(cartItem.getQuantity());
-                    orderItem.setPrice(cartItem.getBook().getPrice());
-                    orderItem.setOrder(order);
+                    OrderItemDto orderItemDto = orderItemMapper.toDto(cartItem);
+                    OrderItem orderItem = orderItemMapper.toEntity(orderItemDto);
```
Comment (olekskov) [inline src/main/java/mate/academy/service/OrderServiceImpl.java]: same
```suggestion
                    OrderItem orderItem = new OrderItem(cartItem);
```

**Review by olekskov — CHANGES_REQUESTED** [review]

**Review by okuzan — APPROVED** [review]

**Pull request merged**